File: core-render-web/base/src/jsMain/kotlin/com/tencent/kuikly/core/render/web/expand/module/KRSseModule.kt
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.kuikly.core.render.web.expand.module

import com.tencent.kuikly.core.render.web.export.KuiklyRenderBaseModule
import com.tencent.kuikly.core.render.web.ktx.KuiklyRenderCallback
import com.tencent.kuikly.core.render.web.ktx.toJSONObjectSafely
import com.tencent.kuikly.core.render.web.nvi.serialization.json.JSONObject

/**
 * SSE（Server-Sent Events）模块 H5 端实现（浏览器 EventSource 封装）。
 * 对应 core 模块 com.tencent.kuikly.core.module.SseModule。
 * 帧投递经 core NotifyModule 全局通道（[KRNotifyModule.dispatchGlobalEvent]）——
 * 该通道为「native 重复推送」的官方机制，不受单次模块回调生命周期限制。
 */
class KRSseModule : KuiklyRenderBaseModule() {

    private var source: dynamic = null

    override fun call(method: String, params: String?, callback: KuiklyRenderCallback?): Any? {
        return when (method) {
            METHOD_CONNECT -> {
                connect(params)
                null
            }
            METHOD_CLOSE -> {
                closeSource()
                null
            }
            else -> super.call(method, params, callback)
        }
    }

    /** 建立 EventSource 连接；帧经 NotifyModule 全局通道投递 */
    private fun connect(params: String?) {
        closeSource() // 同一模块实例同时只允许一条连接
        val url = params?.toJSONObjectSafely()?.optString("url").orEmpty()
        if (url.isEmpty()) {
            postFrame("error", "", "url 为空")
            return
        }
        val es = newEventSource(url)
        es.onopen = { _: dynamic ->
            // 连接建立（含浏览器自动重连成功）：复位 core 侧重连计数并启动看门狗，
            // 否则空闲但健康的连接会一直累积重连次数直到耗尽
            postFrame("open", "", "")
            null
        }
        es.onmessage = { e: dynamic ->
            // lastEventId 透传（Last-Event-ID 断点续传用）
            postFrame("data", (e.data as? String) ?: "", (e.lastEventId as? String) ?: "")
            null
        }
        es.onerror = { _: dynamic ->
            // 浏览器会自动重连；同时上报 error 由上层（看门狗/退避）决策
            postFrame("error", "", "")
            null
        }
        source = es
    }

    /** 模块销毁时关闭 EventSource，避免泄漏 */
    override fun onDestroy() {
        closeSource()
    }

    private fun closeSource() {
        val es = source
        if (es != null) {
            es.close()
        }
        source = null
    }

    /** 经 KRNotifyModule 全局通道投递帧（event/data/id），连接重试后依然可达 */
    private fun postFrame(event: String, data: String, id: String) {
        try {
            KRNotifyModule.dispatchGlobalEvent(
                EVENT_FRAME,
                JSONObject()
                    .put("event", event)
                    .put("data", data)
                    .put("id", id),
            )
        } catch (e: Throwable) {
            // 投递失败静默（页面销毁等场景）
        }
    }

    companion object {
        const val MODULE_NAME = "KRSseModule"
        private const val METHOD_CONNECT = "connect"
        private const val METHOD_CLOSE = "close"
        private const val EVENT_FRAME = "KRSseModule.frame"
    }
}

/** 构造 EventSource（js() 代码体内可直接引用本函数参数） */
private fun newEventSource(url: String): dynamic = js("new EventSource(url)")

File: core-render-ios/Extension/Modules/KRSseModule.m
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "KRSseModule.h"
#import "NSObject+KR.h"

#define SSE_EVENT_FRAME @"KRSseModule.frame"

@interface KRSseModule ()

@property (nonatomic, strong, nullable) NSURLSession *session;
@property (nonatomic, strong, nullable) NSURLSessionDataTask *task;
// 原始字节缓冲：多字节 UTF-8 字符可能跨 chunk，按字节拼到完整帧再解码
@property (nonatomic, strong) NSMutableData *pendingBytes;

@end

@implementation KRSseModule

- (instancetype)init {
    if (self = [super init]) {
        _pendingBytes = [NSMutableData data];
    }
    return self;
}

// core SseModule.connect(url, headers) 分发到本方法
- (id)connect:(NSDictionary *)args {
    [self close:nil];

    NSDictionary *param = [args[KR_PARAM_KEY] hr_stringToDictionary];
    NSString *url = param[@"url"];
    if (![url isKindOfClass:NSString.class] || url.length == 0) {
        [self postFrame:@{ @"event": @"error", @"message": @"url 为空" }];
        return nil;
    }

    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:url]];
    [request setValue:@"text/event-stream" forHTTPHeaderField:@"Accept"];
    NSDictionary *headers = param[@"headers"];
    if ([headers isKindOfClass:NSDictionary.class]) {
        [headers enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
            [request setValue:[NSString stringWithFormat:@"%@", obj] forHTTPHeaderField:key];
        }];
    }

    NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
    config.timeoutIntervalForResource = 0; // SSE 长连接不设整体超时
    self.session = [NSURLSession sessionWithConfiguration:config delegate:self delegateQueue:nil];
    self.task = [self.session dataTaskWithRequest:request];
    [self.task resume];
    return nil;
}

// core SseModule.close() 分发到本方法
- (id)close:(NSDictionary *)args {
    [self.task cancel];
    self.task = nil;
    [self.session invalidateAndCancel];
    self.session = nil;
    self.pendingBytes = [NSMutableData data];
    return nil;
}

// 页面销毁时释放长连接（TDFBaseModule 生命周期钩子）
- (void)invalidate {
    [self close:nil];
    [super invalidate];
}

#pragma mark - NSURLSessionDataDelegate

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data {
    [self.pendingBytes appendData:data];
    // SSE 以 \n\n 分帧（按字节扫描，完整帧再 UTF-8 解码，避免多字节字符跨 chunk 被切断）
    NSData *separator = [@"\n\n" dataUsingEncoding:NSUTF8StringEncoding];
    while (YES) {
        NSRange range = [self.pendingBytes rangeOfData:separator options:0 range:NSMakeRange(0, self.pendingBytes.length)];
        if (range.location == NSNotFound) break;
        NSData *frameData = [self.pendingBytes subdataWithRange:NSMakeRange(0, range.location)];
        [self.pendingBytes replaceBytesInRange:NSMakeRange(0, range.location + range.length) withBytes:NULL length:0];
        NSString *frame = [[NSString alloc] initWithData:frameData encoding:NSUTF8StringEncoding];
        if (frame) [self fireFrame:frame];
    }
}

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error {
    if (error && error.code != NSURLErrorCancelled) {
        [self postFrame:@{ @"event": @"error", @"message": error.localizedDescription ?: @"连接中断" }];
    }
    self.task = nil;
}

- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask
didReceiveResponse:(NSURLResponse *)response completionHandler:(void (^)(NSURLSessionResponseDisposition))completionHandler {
    NSInteger code = [(NSHTTPURLResponse *)response statusCode];
    if (code < 200 || code >= 300) {
        [self postFrame:@{ @"event": @"error", @"message": [NSString stringWithFormat:@"HTTP %ld", (long)code] }];
        completionHandler(NSURLSessionResponseCancel);
        return;
    }
    // 连接建立事件：通知 core 侧复位重连计数并启动看门狗，
    // 否则空闲但健康的连接会一直累积重连次数直到耗尽
    [self postFrame:@{ @"event": @"open" }];
    completionHandler(NSURLSessionResponseAllow);
}

#pragma mark - private

- (void)fireFrame:(NSString *)frame {
    NSMutableString *payload = [NSMutableString string];
    __block NSString *frameId = @"";
    __block BOOL sawComment = NO;
    [frame enumerateLinesUsingBlock:^(NSString *line, BOOL *stop) {
        if ([line hasPrefix:@"data:"]) {
            NSString *content = [line substringFromIndex:5];
            [payload appendString:[content stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceCharacterSet]];
            [payload appendString:@"\n"];
        } else if ([line hasPrefix:@"id:"]) {
            frameId = [[line substringFromIndex:3] stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceCharacterSet];
        } else if ([line hasPrefix:@":"]) {
            sawComment = YES; // 注释行（SSE 规范的心跳/keep-alive）
        }
    }];
    if (payload.length > 0) {
        [self postFrame:@{ @"event": @"data",
                           @"data": [payload stringByTrimmingCharactersInSet:NSCharacterSet.newlineCharacterSet],
                           @"id": frameId }];
    } else if (sawComment) {
        // 纯注释帧（服务端心跳 `: ping`）：透传 heartbeat 喂 core 侧看门狗
        [self postFrame:@{ @"event": @"heartbeat" }];
    }
}

// 帧投递：NSNotificationCenter（core NotifyModule 持久订阅通道，与 SseModule 订阅事件名一致）
- (void)postFrame:(NSDictionary *)frame {
    [[NSNotificationCenter defaultCenter] postNotificationName:SSE_EVENT_FRAME object:nil userInfo:frame];
}

@end
